app/Http/Controllers/Api/OrderController.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+namespace App\Http\Controllers\Api;
+
+use DB;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use App\Contracts\OrderItemRepositoryInterface;
+use App\Contracts\OrderRepositoryInterface;
+use App\Http\Requests\StoreOrder;
+use App\Http\Requests\UpdateOrder;
+use App\Http\Resources\OrderResource;
+
+class OrderController extends ApiController
+{
+    public function __construct(
+        OrderItemRepositoryInterface $orderItemRepository,
+        OrderRepositoryInterface $orderRepository
+    ) {
+        parent::__construct();
+
+        $this->orderItemRepository = $orderItemRepository;
+        $this->orderRepository = $orderRepository;
+    }
+
+    /**
+     * Should return the <RepositoryInterface>::class
+     *
+     * @return string
+     */
+    public function repository()
+    {
+        return OrderRepositoryInterface::class;
+    }
+
+    /**
+     * Should return the <Resource>::class
+     *
+     * @return string
+     */
+    public function resource()
+    {
+        return OrderResource::class;
+    }
+
+    /**
+     * Should return <StoreRequest>::class
+     *
+     * @return string
+     */
+    public function storeRequest()
+    {
+        return StoreOrder::class;
+
+        // TODO Due to the parent class, the Request cannot be type-hinted on
+        // this method. How best to access the Request object, then?
+    }
+
+    /**
+     * Should return <UpdateRequest>::class
+     *
+     * @return string
+     */
+    public function updateRequest()
+    {
+        return UpdateOrder::class;
+    }
+
+    /**
+     * Create an Order, which requires adding one or more Items to the Order.
+     *
+     * @param Request $request
+     * @return Response
+     */
+    public function storeOrder(StoreOrder $request)
```

Comment:
1.) The method you propose to solve the unauthorized manipulation seems to be acceptable IMO, However I'm not so sure that I am the most qualified in this particular area to say if its the BEST method.

2.) I'm confused on the TODO issue tbh ... Not sure what the requirement vs complication is as its not entirely clear.

3.) Suggestion on Request Class naming conventions appending `Request` ... ie: `StoreOrderRequest` . Whats your take on this ?

Reply:
1.) Very good. I've spent enough time reasoning through it that I'm comfortable running with it if and until a better approach emerges.

2.) Sorry, I should have been more explicit regarding the TODO.

Fundamentally, the problem I encountered is that the pattern we're employing currently prevents us from type-hinting the Request object on the various controller methods.

If you look at the methods I added that don't conform to our existing pattern, `storeOrder()`, `addItems()` and `removeItems()`, they all deviate in that they type-hint a Request object. Basically, we need access to the Request object to be able to retrieve and manipulate the POST inputs in a way that is more complex than has been required elsewhere in the code-base to date.

Basically, I created the `storeOrder(StoreOrder $request)` method to use in place of the "built-in" `storeRequest()` method, because I need access to the Request object (via the `$request` variable as seen in the method signature). The problem is that we can't type-hint like this on the `storeRequest()` method, e.g., `storeRequest(StoreOrder $request)`, because doing so results in `Declaration must be compatible with ApiController->storeRequest()`.

In short, if you can find a way to achieve what I've done in those three methods while adhering to the current pattern, I'd be very curious to see how you do it.

3.) I like it. I'm happy to rename them.

Reply:
Regarding the TODO, if you have a working solution right this minute i say lets move forward, but i would like to make note to revisit this once i'm back at it. I think there could be better approach to this but am not really sure what that is atm. Instead of tying this PR up lets get it merged in.

Could you please open a issue regarding this topic so that it doesn't get lost ?

Reply:
I concur! Yes, I'll create an Issue about it right now, so we don't forget.